source/mbed_trace.c

```diff
@@ -213,8 +231,18 @@ void mbed_trace_free(void)
 }
 static void mbed_trace_realloc(char **buffer, int *length_ptr, int new_length)
 {
-    MBED_TRACE_MEM_FREE(*buffer);
-    *buffer  = MBED_TRACE_MEM_ALLOC(new_length);
+    void *new_buffer = NULL;
+    if (new_length != 0) {
+        new_buffer = MBED_TRACE_MEM_ALLOC(new_length);
+        // TODO: This may fail
+        if (*buffer != NULL) {
+            memcpy(new_buffer, *buffer, (*length_ptr > new_length)?new_length:*length_ptr);
+        }
+    }
+    if (*buffer != NULL) {
+        MBED_TRACE_MEM_FREE(*buffer);
+    }
+    *buffer = new_buffer;
     *length_ptr = new_length;
```

Comment:
Perhaps this function could just be replaced with a `MBED_TRACE_MEM_REALLOC` macro? Set the default value of it to std realloc, and default back to this if the macro isn't defined? Could avoid this extra copy most of the time.

Reply:
That makes sense.

Although `mbed-trace` had a realloc function that forced a free & allocation (although not guaranteeing a new location) prior to this PR.
Do you know why and how using the real `realloc` would impact it?